src/vs/editor/common/model/tokenizationTextModelPart.ts

```diff
@@ -44,7 +49,8 @@ export class TokenizationTextModelPart extends TextModelPart implements ITokeniz
 	private readonly _onDidChangeTokens: Emitter<IModelTokensChangedEvent> = this._register(new Emitter<IModelTokensChangedEvent>());
 	public readonly onDidChangeTokens: Event<IModelTokensChangedEvent> = this._onDidChangeTokens.event;
 
-	private readonly grammarTokens = this._register(new GrammarTokens(this._languageService.languageIdCodec, this._textModel, () => this._languageId, this._attachedViews));
+	private grammarTokens: GrammarTokens | undefined;
+	private treeSitterTokens: TreeSitterTokens | undefined;
```

Comment:
I'd suggest to not have a separate field `treeSitterTokens`, but have a tree sitter implemenation of `GrammarTokens` (and introducing an abstract class).

Reply:
Sounds good, I will address in an upcoming commit.